ITV2: disallow duplicate dynamic templates

andreidan opened the pull request:
Dynamic templates can contain multiple templates with the same name. This
commit changes the way V2 index templates mappings are resolved to add
deduplication of dynamic templates when resolving the mappings, by having
the last dynamic templates override the ones defined with the same name
earlier in the `dynamic_templates` array.

Relates to #53101
Relates to #53326
Relates to #28988

andreidan committed:
ITV2: disallow duplicate dynamic templates

Dynamic templates can contain multiple templates with the same name. This
commit changes the way V2 index templates mappings are resolved to add
deduplication of dynamic templates when resolving the mappings, by having
the last dynamic templates override the ones defined with the same name
earlier in the `dynamic_templates` array.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Indices APIs)

dakrone commented:
@andreidan can you add a test that component templates merge their dynamic templates correctly? I think because we merge the XContent maps prior to the dedupe it is possible to end up with an invalid dynamic template if you have two component templates that each define a dynamic template with the same name but different configurations that, when merged, make the template invalid.

andreidan commented:
@dakrone thanks for the suggestion on the component templates merging. Great shout on testing that as you were right and we are merging the Maps so, even though we dedup the key names (ie. there will only be one `dynamic_template` with a particular name) the bodies are merged.

andreidan committed:
Filter duplicate dynamic templates

This filters duplicate dynamic templates when merging different component
templates that specify the same dynamic template and when merging the
mappings specified in the request with the ones in the index template.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into dedup_dynamic_templates

andreidan committed:
Fix line length

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -576,6 +577,18 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
                 Map<String, Object> innerTemplateNonProperties = new HashMap<>(innerTemplateMapping);
                 Map<String, Object> maybeProperties = (Map<String, Object>) innerTemplateNonProperties.remove("properties");
 
+                List<Map<String, Object>> innerTemplateDynamicTemplates = (List<Map<String, Object>>) innerTemplateNonProperties.get(
+                    "dynamic_templates");
+                List<Map<String, Object>> previouslySeenDynamicTemplates =
+                    (List<Map<String, Object>>) nonProperties.get("dynamic_templates");
+                // we want "later" defined dynamic templates to override the previously seen ones with the same name
+                List<Map<String, Object>> filteredDefaultDynamicTemplates =
+                    removeAll(previouslySeenDynamicTemplates, innerTemplateDynamicTemplates);
+
+                if (filteredDefaultDynamicTemplates != previouslySeenDynamicTemplates) {
+                    nonProperties.put("dynamic_templates", filteredDefaultDynamicTemplates);
+                }
```
I think we can encapsulate this logic into a method to make it a bit clearer what's going on, perhaps something like:

```
nonProperties = removeDuplicatedDynamicTemplates(nonProperties, innerTemplateNonProperties);
```

Then the same method can be used below (where these lines are almost identical except for the maps passed in).

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -598,11 +623,78 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
+     * Returns:
+     *     [ {"key2" : {}} ]
+     */
+    private static List<Map<String, Object>> removeAll(List<Map<String, Object>> first, List<Map<String, Object>> second) {
```
Maybe call this `removeOverlapping` to make it a bit clearer what it's doing?

dakrone reviewed (changes requested):
I left a couple of comments similar to what we talked about regarding naming, thanks for doing this Andrei!

andreidan committed:
Extract method and re-name removeAll

andreidan committed:
Update doc

dakrone reviewed (approved):
LGTM, thanks for working on this Andrei!
